### PR TITLE
Issue/731 randomization

### PR DESCRIPTION
**Pull request by pjljvandelaar**
related to #731

### CODE REVIEW

**Comment by pjljvandelaar** [comment]
@keremispirli what is the performance impact?

**Comment by keremispirli** [comment]
@pjljvandelaar The branch still has conflicts; could you merge from develop? Then I'll start a benchmark to see performance.

**Comment by pjljvandelaar** [comment]
@keremispirli why is semaphoreci absent?

**Comment by keremispirli** [comment]
Possibly another webhook glitch; let's see if we have it when `develop` is merged.

**Comment by pjljvandelaar** [comment]
Seems I missed some push or pop
....

**Comment by pjljvandelaar** [comment]
Indeed missed hlint issue...
Why is semaphoreci absent as checker?

**Comment by pjljvandelaar** [comment]
Fine on appveyor and semaphoreci!

**Comment by keremispirli** [comment]
The webhook looks successful, I redelivered it twice but for some reason SemphoreCI continued ignoring this branch.
Then I discovered a tiny (+) sign next to "Branches" title on the project page of TorXakis, which I could enable this branch. Now it's running.

**Comment by pjljvandelaar** [comment]
No failing tests @ appveyor... just out of time.
Should we 
1. just accept (earlier commits did succeed)
2. rerun until appveyor makes it in time
3. move back to IncrementChoice which is faster for customer orders (smaller strings)
4. change random string generation: instead of intervals based on [0,3,9,27] use for example [0,2,4,8,16]?

**Comment by dnadales** [comment]
I just restarted appveyor. I guess this is another reason for biting the bullet and replacing the current compiler. We might gain a lot of time by avoiding long compilation times. I'm going to benchmark how much time we will gain after I'm done with the cleanup (Friday or Monday).

**Comment by keremispirli** [comment]
Meanwhile I'm running the benchmarks on this version. Stay tuned.

**Comment by pjljvandelaar** [comment]
@keremispirli I am mainly interested in the difference between IncrementBins and IncrementChoice on this branch: the most equal comparison imho.

**Comment by pjljvandelaar** [comment]
@capitanbatata seems something wrong at appveyor: the others are canceled, yet this build doesn't start ;-(

**Comment by dnadales** [comment]
I restarted it again. Let's see...

**Comment by dnadales** [comment]
Ok, now it started right away...

**Review by keremispirli — CHANGES_REQUESTED** [review]
@pjljvandelaar With this version, I cannot assign `param_Randomization` to `IncrementBins`.

```
TXS >> param param_Randomization IncrementBins
TXS >>  No parameter: param_Randomization
```
Note that the error message is misleading; it says "No parameter" but actually error happens because IncrementBins is not recognized.

```
TXS >> param param_Randomization TrueBinsa
TXS >>  No parameter: param_Randomization
```

It does work for different randomization parameters though:
```
TXS >> param param_Randomization
TXS >>  param_Randomization = IncrementChoice
TXS >> param param_Randomization TrueBins
TXS >>  param_Randomization = TrueBins
```

**Comment by keremispirli** [comment]
Another issue: There's no guarantee that the `.torxakis.yaml` file at the root of the repo will be used for configuration. In order to set default randomization to `IncrementBins`, L60 in `\sys\solve\src\Solve\Params.hs` should also be changed to `IncrementBins`.

**Comment by pjljvandelaar** [comment]
Appveyor again stopped after 1 hour. 
I noticed it did not use the stack cache: everything was build from scratch...

File: sys/solve/src/Solve/Params.hs [hunk 1]
```diff
@@ -61,7 +61,8 @@ initParams  =  Map.fromList $ map ( \(x,y,z) -> (x,(y,z)) )
                                                                           , \s ->    (s == show No)
                                                                                   || (s == show Partition)
                                                                                   || (s == show TrueBins)
-                                                                                  || (s == show IncrementChoice) )
+                                                                                  || (s == show IncrementChoice) 
+                                                                                  || (s == show IncrementBins) )
```
Comment (pjljvandelaar) [inline sys/solve/src/Solve/Params.hs:65]: Added IncrementBins to check function, so don't understand why not accepted. Will investigate tomorrow.

**Comment by pjljvandelaar** [comment]
@keremispirli I can't reproduce
```
TXS >>  TorXakis :: Model-Based Testing

TXS >>  txsserver starting: "NLYBSTQVP4NB77Y.code1.emi.philips.com" : 63044
TXS >>  Solver "z3" initialized : Z3 [4.8.0 - build hashcode a12af4a619a0]
TXS >>  TxsCore initialized
TXS >>  input files parsed:
TXS >> param
TXS >>  param_ImpRel = IOCO
TXS >>  param_IncrementChoice_IntPower = 4
TXS >>  param_IncrementChoice_IntRange = 65536
TXS >>  param_IncrementChoice_MaxGeneratedStringLength = 10
TXS >>  param_InputCompletion = ANGELIC
TXS >>  param_RandSolve_IntHalf = 1000
TXS >>  param_RandSolve_IntNum = 5
TXS >>  param_Randomization = IncrementBins
TXS >>  param_Sim_inputEager = 0
TXS >>  param_Test_inputEager = 3
TXS >>  param_TrueBins_Next = Linear
TXS >>  param_TrueBins_NrOfBins = 10
TXS >>  param_TrueBins_StringLength = 6
TXS >>  param_TrueBins_StringMode = Regex
TXS >>  param_max_rand_depth = 4
TXS >>  param_Sim_deltaTime = 2000
TXS >>  param_Sim_ioTime = 10
TXS >>  param_Sut_deltaTime = 2000
TXS >>  param_Sut_ioTime = 10
TXS >> param param_Randomization TrueBins
TXS >>  param_Randomization = TrueBins
TXS >> param param_Randomization IncrementBins
TXS >>  param_Randomization = IncrementBins
TXS >>
```

Did you only build and not install?

File: sys/solve/src/Solve/Params.hs (hunk 1 above)
Comment (pjljvandelaar) [reply sys/solve/src/Solve/Params.hs:65]: Made remark yesterday, yet review was not started.....

**Comment by pjljvandelaar** [comment]
Do we want duplication in test code?
It turns out that I also have to change the test code, now that I changed the default value: it is not derived from the original code.
Will send commit soon.

**Comment by keremispirli** [comment]
Strange. I run the script which runs `stack install`. I had still suspected that, so tried to reinstall but stack just copied the same executables. I'll try again with this version.

**Comment by pjljvandelaar** [comment]
After 28:15 minutes the actual tests start @ appveyor.
Semaphoreci is finished in 25:57 minutes...

I think the new randomization solution is slower, yet better distributed.
The question is how further?
Let's discuss on Monday!

**Comment by keremispirli** [comment]
I'll run a no-test build with lts 11.5 so that it succeeds and caches `.stack-root`. That should cut some of the build time.

**Comment by pjljvandelaar** [comment]
@keremispirli Great!

**Comment by keremispirli** [comment]
Looks like everything was already cached. So this failing build was merely a cache-miss. I restarted it; now it fetched `.stack-root` and is building with it.

**Comment by pjljvandelaar** [comment]
Just in Time: 59 min and 15 seconds!

**Comment by pjljvandelaar** [comment]
@keremispirli Is your review solved?
If so, please accept!

**Comment by keremispirli** [comment]
It's solved and benchmark of IncrementBins is also finished, but I'm still waiting for benchmark of IncrementChoice. Should I accept regardless?

**Comment by pjljvandelaar** [comment]
No, let's discuss on Monday!

**Comment by keremispirli** [comment]
Here are the benchmark results: [Randomization_IncChoice_vs_IncBins.xlsx](https://github.com/TorXakis/TorXakis/files/2132572/Randomization_IncChoice_vs_IncBins.xlsx)

Key outcomes:
 - "_Hiding/match Int_" is 526.6 % faster
 - "_Real World/Multiple Conrol Loops Stepper_" is 58.68 % faster
 - "_Real World/Customers and Orders_" is 73.5% **slower**
 --> This is because random strings generated by IncrementBins are longer than the ones from IncrementChoice.
 - All "_Sequence/sequence with a 10 integer channel_" cases are ~50 % **slower**
--> Because there are more bins tried.

**Review by keremispirli — APPROVED** [review]

**Pull request merged**